### PR TITLE
8315098: Improve URLEncodeDecode microbenchmark

### PR DESCRIPTION
**Pull request by cl4es**
The `URLEncodeDecode` microbenchmark accidentally generates strings with a lot of `'\u0000'` chars, heavily skewing towards strings that need to be encoded in a rather unrealistic what. To be more realistic the benchmark should test a mix of inputs.

This patch fixes these inadvertent cases, and sets up the benchmark for a healthier mix by default - adding controls to allow testing some mixed scenarios.

#15354 explore a few optimizations to `URLEncoder`, but due the nature of this microbenchmark a trivial fast-path scan for chars that need no encoding shows underwhelming results. With the modifications to this benchmark then a simple fast-path to `URLEncode.encode` shows a decent win when some or all the inputs remain unchanged:

```
Name                           (encodeChars) (maxLength) (unchanged) Cnt  Base   Error   Test   Error  Unit  Diff%
URLEncodeDecode.testDecodeUTF8             6        1024           0  15 3,307 ± 0,507  3,010 ± 0,048 ms/op   9,0% (p = 0,030 )
URLEncodeDecode.testDecodeUTF8             6        1024          75  15 2,296 ± 0,003  2,313 ± 0,017 ms/op  -0,7% (p = 0,001*)
URLEncodeDecode.testDecodeUTF8             6        1024         100  15 0,812 ± 0,010  0,819 ± 0,017 ms/op  -0,8% (p = 0,201 )
URLEncodeDecode.testDecodeUTF8            35        1024           0  15 6,909 ± 0,065  7,192 ± 0,415 ms/op  -4,1% (p = 0,014 )
URLEncodeDecode.testDecodeUTF8            35        1024          75  15 3,346 ± 0,206  3,320 ± 0,270 ms/op   0,8% (p = 0,753 )
URLEncodeDecode.testDecodeUTF8            35        1024         100  15 0,794 ± 0,034  0,818 ± 0,015 ms/op  -3,0% (p = 0,016 )
URLEncodeDecode.testEncodeUTF8             6        1024           0  15 2,434 ± 0,019  2,579 ± 0,120 ms/op  -6,0% (p = 0,000*)
URLEncodeDecode.testEncodeUTF8             6        1024          75  15 1,764 ± 0,014  0,937 ± 0,012 ms/op  46,9% (p = 0,000*)
URLEncodeDecode.testEncodeUTF8             6        1024         100  15 1,227 ± 0,008  0,401 ± 0,001 ms/op  67,4% (p = 0,000*)
URLEncodeDecode.testEncodeUTF8            35        1024           0  15 6,177 ± 0,062  6,057 ± 0,199 ms/op   1,9% (p = 0,029 )
URLEncodeDecode.testEncodeUTF8            35        1024          75  15 2,716 ± 0,023  1,876 ± 0,012 ms/op  30,9% (p = 0,000*)
URLEncodeDecode.testEncodeUTF8            35        1024         100  15 1,220 ± 0,003  0,401 ± 0,001 ms/op  67,2% (p = 0,000*)
```

A potential future improvement would be to extend test data with varying amounts of surrogate pairs, e.g. Chinese or emojis

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315098](https://bugs.openjdk.org/browse/JDK-8315098): Improve URLEncodeDecode microbenchmark (**Enhancement** - P4)

### Reviewers
 * [Eric Caspole](https://openjdk.org/census#ecaspole) (@ericcaspole - Committer) ⚠️ Review applies to [0d28ef05](https://git.openjdk.org/jdk/pull/15448/files/0d28ef05f05e4056a19f493d1c14d9ac9a5fd5f3)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15448/head:pull/15448` \
`$ git checkout pull/15448`

Update a local copy of the PR: \
`$ git checkout pull/15448` \
`$ git pull https://git.openjdk.org/jdk.git pull/15448/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15448`

View PR using the GUI difftool: \
`$ git pr show -t 15448`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15448.diff">https://git.openjdk.org/jdk/pull/15448.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15448#issuecomment-1695725307)

### CODE REVIEW

**Review by dfuch — APPROVED** [review]
Thanks for improving this microbenchmark Claes. Changes look good to me. Using 1024 strings  should ensure that at least some of them have some characters that need to be encoded/decoded.

**Comment by cl4es** [comment]
> Idle musing: I wonder if it would be useful to print on System.out how many strings actually have no encoded chars? Just to get a rough feeling on whether the percentage values were actually respected.

Good call, since I had gotten the distribution wrong for decode with `unchanged` set to a value like 75 by recalculating `needDecoding` for every char. Also improved the precision for `unchanged == 0`.

**Review by dfuch — APPROVED** [review]

File: test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java [hunk 1]
```diff
@@ -51,78 +52,140 @@
...
+        for (char c = 'a'; c <= 'z'; c++) {
+            tokens[n++] = c;
         }
         tokens[n++] = '-';
         tokens[n++] = '_';
         tokens[n++] = '.';
-        tokens[n++] = '*';
+        tokens[n] = '*';
 
-        Random r = new Random(mySeed);
```
Comment (franz1981) [inline test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java:86]: SplittableRandom allow to have perfectly reproducible sequences

File: test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java [hunk 2]
```diff
@@ -51,78 +52,140 @@
...
+        if (r.nextInt(100) < 15) {
+            sb.append('+'); // exercise '+' -> ' ' decoding paths.
+        } else {
+            sb.append("%").append(tokens[r.nextInt(16)]).append(tokens[r.nextInt(16)]);
+        }
     }
 
     @Benchmark
     public void testEncodeUTF8(Blackhole bh) throws UnsupportedEncodingException {
         for (String s : testStringsEncode) {
```
Comment (franz1981) [inline test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java:180]: The consume is pretty cheap with compiler assisted blackholes, but can still overall be too noisy with small 's'

Worth running perfasm and see what it says and/or reading put of the encoded value and extract some property to accumulate in the for loop. And see if is less costy.

Another point instead is related loop unrolling/strip mining; safepoint polls can create assitionalrnandom noise depending how much loop unrolling is excercized (Vs the cost of encoding): it worth to enforce a specific unrolling level or disabling it as a whole, too, to get more stable results.

File: test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java (hunk 1 above)
Comment (cl4es) [reply test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java:86]: Care to elaborate? `Random` with an explicit seed should be more than enough to have perfectly reproducible sequences for our case (`SplittableRandom` uses a different algorithm by default with a longer period, but I can't see how this'd be relevant to this case.)

```
 * If two instances of {@code Random} are created with the same
 * seed, and the same sequence of method calls is made for each, they
 * will generate and return identical sequences of numbers.
```

Comment (franz1981) [reply test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java:86]: Even with the same seed only with large numbers you get statistically a similar distribution,m (not the same values!), but splittable random is actually producing the same exact sequences. You care about confusing branch predictors here I suppose (or the JIT, maybe too) but not to add noise or variance with fewer samples nor you are interested to the security part of it (cause random use the chaos generated securely by the OS/HW iirc).

Comment (franz1981) [reply test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java:86]: I have to check in my notes why I used splittable rmd and will update this thread :)

File: test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java (hunk 2 above)
Comment (cl4es) [reply test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java:180]: I think these are all valid concerns and it's a worthwhile endeavor to discuss and try and improve how we benchmark loops (or more specifically variable inputs). However I think this is out of scope for this improvement - which is mainly about correctness of data and making sure we're testing the right thing - and ought to be discussed and experimented with separately. If you don't strongly disagree I will integrate this so we can move ahead with other work in this area (#15354 for starters)

**Pull request closed without merging**